x/slashing/src/keeper.rs

```diff
@@ -87,6 +103,211 @@ impl<SK: StoreKey, PSK: ParamsSubspaceKey, SSK: SlashingStakingKeeper<SK, M>, M:
         self.slashing_params_keeper.set(ctx, genesis.params);
     }
 
+    pub fn handle_validator_signature<DB: Database>(
+        &self,
+        ctx: &mut BlockContext<'_, DB, SK>,
+        addr_bytes: prost::bytes::Bytes,
+        power: u32,
+        signed: bool,
+    ) {
+        let height = ctx.height();
+
+        // fetch the validator public key
+        let cons_addr = ConsAddress::try_from(Vec::from(addr_bytes)).unwrap();
+        self.get_pub_key(ctx, &cons_addr)
+            .expect("validator consensus address not found");
```

Comment:
This method used in `block_end` so @dmvict  omited it

Reply:
Yes, the method is used by `begin_block` method and should be infallible. I think it is better to save this behavior because the method is used only by `begin_block`.

Reply:
I don't think you have enough context to panic at this point. The method accepts an arbitrary validator which may or may not exist, you can't tell from inside the method whether it is expected to exist.

Reply:
Updated.